File: src/app/nutrition/micronutrients.py
"""Registro canonico de micronutrientes.

Unica fuente de verdad para claves, unidades, RDA (adulto de referencia) y
limite superior seguro (upper_limit). Se expone via
`GET /diet/micronutrients` para que el frontend no duplique la tabla, y lo
usa `app.nutrition.mapping` para saber en que unidad debe quedar cada valor
convertido desde Open Food Facts.

Valores de referencia tomados de las Dietary Reference Intakes (NIH/EFSA)
para un adulto promedio; son un default razonable, no una recomendacion
clinica personalizada. `upper_limit` queda en None cuando no hay un limite
superior tolerable establecido.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class MicronutrientDefinition:
    key: str
    label: str
    unit: str
    rda: float
    upper_limit: float | None


MICRONUTRIENTS: tuple[MicronutrientDefinition, ...] = (
    MicronutrientDefinition("vitamin_a_ug", "Vitamina A", "ug", 900.0, 3000.0),
    MicronutrientDefinition("vitamin_c_mg", "Vitamina C", "mg", 90.0, 2000.0),
    MicronutrientDefinition("vitamin_d_ug", "Vitamina D", "ug", 20.0, 100.0),
    MicronutrientDefinition("vitamin_e_mg", "Vitamina E", "mg", 15.0, 1000.0),
    MicronutrientDefinition("vitamin_k_ug", "Vitamina K", "ug", 120.0, None),
    MicronutrientDefinition("thiamin_mg", "Tiamina (B1)", "mg", 1.2, None),
    MicronutrientDefinition("riboflavin_mg", "Riboflavina (B2)", "mg", 1.3, None),
    MicronutrientDefinition("niacin_mg", "Niacina (B3)", "mg", 16.0, 35.0),
    MicronutrientDefinition("vitamin_b6_mg", "Vitamina B6", "mg", 1.7, 100.0),
    MicronutrientDefinition("folate_ug", "Folato", "ug", 400.0, 1000.0),
    MicronutrientDefinition("vitamin_b12_ug", "Vitamina B12", "ug", 2.4, None),
    MicronutrientDefinition("calcium_mg", "Calcio", "mg", 1300.0, 2500.0),
    MicronutrientDefinition("iron_mg", "Hierro", "mg", 18.0, 45.0),
    # El UL de 350 mg del magnesio aplica SOLO al magnesio suplementario, no al
    # dietario (NIH ODS). Dejarlo aqui marcaria como "excedido" a cualquiera que
    # cumpla su RDA de 420 mg comiendo normal: falsa alarma diaria para todos.
    MicronutrientDefinition("magnesium_mg", "Magnesio", "mg", 420.0, None),
    MicronutrientDefinition("phosphorus_mg", "Fosforo", "mg", 1250.0, 4000.0),
    MicronutrientDefinition("potassium_mg", "Potasio", "mg", 4700.0, None),
    MicronutrientDefinition("zinc_mg", "Zinc", "mg", 11.0, 40.0),
    MicronutrientDefinition("copper_mg", "Cobre", "mg", 0.9, 10.0),
    MicronutrientDefinition("manganese_mg", "Manganeso", "mg", 2.3, 11.0),
    MicronutrientDefinition("selenium_ug", "Selenio", "ug", 55.0, 400.0),
    MicronutrientDefinition("iodine_ug", "Yodo", "ug", 150.0, 1100.0),
)

MICRONUTRIENTS_BY_KEY: dict[str, MicronutrientDefinition] = {
    item.key: item for item in MICRONUTRIENTS
}

MICRONUTRIENT_KEYS: frozenset[str] = frozenset(MICRONUTRIENTS_BY_KEY)


def get_micronutrient(key: str) -> MicronutrientDefinition | None:
    return MICRONUTRIENTS_BY_KEY.get(key)

File: frontend/src/lib/nutrition/micronutrients.ts
/**
 * Registro canonico de micronutrientes — espejo en TS del futuro
 * `src/app/nutrition/micronutrients.py` (backend, doc `modulo-dieta.md` #3).
 *
 * Mientras el backend expone `GET /diet/micronutrients`, este archivo es la
 * única fuente de verdad en el frontend para que el parser de etiquetas y la
 * UI (barras `MicroBar`, formulario de captura) no inventen cada uno su
 * propia lista. Cuando el endpoint exista, este archivo se puede reemplazar
 * por datos traidos del servidor sin cambiar el tipo `MicronutrientEntry`.
 *
 * Los valores de `rda` (referencia diaria) y `upperLimit` (límite superior
 * tolerable) siguen la tabla de "Daily Values" de FDA/NIH-ODS para adultos,
 * que es la base que también usan las resoluciones colombianas de rotulado
 * (810 de 2021) para calcular %VD. El documento de especificacion no fija
 * números exactos, así que estos son ajustables sin romper el contrato del
 * tipo si el backend define otros.
 *
 * IMPORTANTE — esto NO es lo mismo que la "trampa" del cliente de Open Food
 * Facts (doc #3 y #4): esa conversion (x1000 / x1 000 000) aplica a los
 * valores que devuelve la API de OFF, que vienen en gramos. Las etiquetas
 * impresas que procesa `labelParser.ts` ya declaran mg/µg directamente, así
 * que aquí no hace falta ninguna conversion de unidades.
 */

import type { MicronutrientEntry, MicronutrientKey } from "./types";

export const MICRONUTRIENTS: readonly MicronutrientEntry[] = [
  { key: "vitamin_a_ug", label: "Vitamina A", unit: "ug", rda: 900, upperLimit: 3000 },
  { key: "vitamin_c_mg", label: "Vitamina C", unit: "mg", rda: 90, upperLimit: 2000 },
  { key: "vitamin_d_ug", label: "Vitamina D", unit: "ug", rda: 20, upperLimit: 100 },
  { key: "vitamin_e_mg", label: "Vitamina E", unit: "mg", rda: 15, upperLimit: 1000 },
  { key: "vitamin_k_ug", label: "Vitamina K", unit: "ug", rda: 120, upperLimit: null },
  { key: "thiamin_mg", label: "Tiamina (B1)", unit: "mg", rda: 1.2, upperLimit: null },
  { key: "riboflavin_mg", label: "Riboflavina (B2)", unit: "mg", rda: 1.3, upperLimit: null },
  { key: "niacin_mg", label: "Niacina (B3)", unit: "mg", rda: 16, upperLimit: 35 },
  { key: "vitamin_b6_mg", label: "Vitamina B6", unit: "mg", rda: 1.7, upperLimit: 100 },
  { key: "folate_ug", label: "Folato", unit: "ug", rda: 400, upperLimit: 1000 },
  { key: "vitamin_b12_ug", label: "Vitamina B12", unit: "ug", rda: 2.4, upperLimit: null },
  { key: "calcium_mg", label: "Calcio", unit: "mg", rda: 1300, upperLimit: 2500 },
  { key: "iron_mg", label: "Hierro", unit: "mg", rda: 18, upperLimit: 45 },
  // El UL de 350 mg del magnesio aplica SOLO al magnesio suplementario, no al
  // dietario (NIH ODS). Dejarlo aquí marcaria como "excedido" a cualquiera que
  // cumpla su RDA de 420 mg comiendo normal: falsa alarma diaria para todos.
  { key: "magnesium_mg", label: "Magnesio", unit: "mg", rda: 420, upperLimit: null },
  { key: "phosphorus_mg", label: "Fosforo", unit: "mg", rda: 1250, upperLimit: 4000 },
  { key: "potassium_mg", label: "Potasio", unit: "mg", rda: 4700, upperLimit: null },
  { key: "zinc_mg", label: "Zinc", unit: "mg", rda: 11, upperLimit: 40 },
  { key: "copper_mg", label: "Cobre", unit: "mg", rda: 0.9, upperLimit: 10 },
  { key: "manganese_mg", label: "Manganeso", unit: "mg", rda: 2.3, upperLimit: 11 },
  { key: "selenium_ug", label: "Selenio", unit: "ug", rda: 55, upperLimit: 400 },
  { key: "iodine_ug", label: "Yodo", unit: "ug", rda: 150, upperLimit: 1100 },
] as const;

export const MICRONUTRIENT_BY_KEY: ReadonlyMap<MicronutrientKey, MicronutrientEntry> = new Map(
  MICRONUTRIENTS.map((entry) => [entry.key, entry]),
);

/**
 * Alias en español (normalizados: minusculas, sin tildes) que el parser de
 * etiquetas reconoce para cada micronutriente. Incluye variantes comunes de
 * OCR y nombres alternos usados en etiquetas colombianas reales.
 * `normalizeForMatch` en `labelParser.ts` produce el mismo formato de texto,
 * así que las claves de este mapa deben quedar ya sin tildes.
 */
export const MICRONUTRIENT_ALIASES: ReadonlyMap<string, MicronutrientKey> = new Map([
  ["vitamina a", "vitamin_a_ug"],
  ["vitamina c", "vitamin_c_mg"],
  ["acido ascorbico", "vitamin_c_mg"],
  ["vitamina d", "vitamin_d_ug"],
  ["vitamina e", "vitamin_e_mg"],
  ["vitamina k", "vitamin_k_ug"],
  ["vitamina b1", "thiamin_mg"],
  ["tiamina", "thiamin_mg"],
  ["vitamina b2", "riboflavin_mg"],
  ["riboflavina", "riboflavin_mg"],
  ["vitamina b3", "niacin_mg"],
  ["niacina", "niacin_mg"],
  ["vitamina b6", "vitamin_b6_mg"],
  ["piridoxina", "vitamin_b6_mg"],
  ["folato", "folate_ug"],
  ["acido folico", "folate_ug"],
  ["vitamina b9", "folate_ug"],
  ["vitamina b12", "vitamin_b12_ug"],
  ["cobalamina", "vitamin_b12_ug"],
  ["calcio", "calcium_mg"],
  ["hierro", "iron_mg"],
  ["magnesio", "magnesium_mg"],
  ["fosforo", "phosphorus_mg"],
  ["potasio", "potassium_mg"],
  ["zinc", "zinc_mg"],
  ["cinc", "zinc_mg"],
  ["cobre", "copper_mg"],
  ["manganeso", "manganese_mg"],
  ["selenio", "selenium_ug"],
  ["yodo", "iodine_ug"],
]);
